nightingale_navigation/hallway_planner/src/hallway_locator.cpp

```diff
@@ -0,0 +1,137 @@
+        };
+        std::vector<std::vector<Line>> groups;
+        for(const auto &l : lines)
+        {
+            bool group_found = false;
+            for (auto &group : groups)
+            {
+                if (similar(l,group[0]))
+                {
+                    group.push_back(l);
+                    group_found = true;
+                    break;
+                }
+            }
+            if (!group_found)
+            {
+                std::vector new_group = {l};
+                groups.push_back(std::move(new_group));
+            }
+        }
+        assert(std::none_of(groups.cbegin(),groups.cend(),[](const auto &v){return v.empty();}));
+        assert(std::accumulate(groups.cbegin(),groups.cend(), 0,[](const auto &sum, const auto &v){return sum + v.size();}) == lines.size());
+        assert(groups.size()>0 && groups.size()<=lines.size());
+        std::deque<Line> filtered_lines;
+        std::transform(groups.cbegin(),groups.cend(),std::back_inserter(filtered_lines),[](const auto &v){return v.at(0);});
+        return filtered_lines;
+    }();
+    ROS_INFO_STREAM("HallwayLocator filtered " << lines.size() << " lines down to " << filtered_lines.size() << " lines");
+    const auto hallways = [&filtered_lines,&config]() {
+        // 2 lines are said to be very similar if they are close in hough space
+        const auto dissimilarity = [](const Line &a, const Line &b) {
+            const Point c(a.rho(),a.theta());
+            const Point d(b.rho(),b.theta());
+            return geometry_primitives::euclidean_distance(c,d);
+        };
+        std::vector<Hallway> ret;
+        while (filtered_lines.size() > 1)
+        {
+            const auto cur = filtered_lines.at(0);
+            const auto most_similar = std::min_element(filtered_lines.cbegin()+1,filtered_lines.cend(),[&cur,&dissimilarity](const Line &a,const Line &b){return dissimilarity(a,cur) < dissimilarity(b,cur);});
+            // sometimes, the detector detects lone lines ie. it detects only half a hallway
+            // in these cases, we ignore the line
+            constexpr auto experimental_threshold  = 100;
+            if (dissimilarity(cur, *most_similar) > experimental_threshold)
+            {
+                ROS_INFO("HallwayLocator discarding unmatched hallway line");
+                filtered_lines.pop_front();
+            }
+            else
+            {
+                ret.emplace_back(cur,*most_similar);
+                filtered_lines.erase(most_similar);
+                filtered_lines.pop_front();
+            }
+        }
+        return ret;
+    }();
+    ROS_INFO_STREAM("hallway_planner::hallway_locator::locate_hallways found " << hallways.size() << " hallways."); 
+    cv::Mat hallway_map(costmap.size(),costmap.type(),HallwayLocatorConfig::NO_HALLWAY);
+    for (const auto &hallway : hallways)
```

Comment:
Is this where you will assign the costs inside the hallway? So far it seems like you are just creating the hallway space correct?